src/rust/engine/fs/src/store.rs

```diff
@@ -256,54 +279,201 @@ mod local {
     //  1. They may have different lifetimes.
     //  2. It's nice to know whether we should be able to parse something as a proto.
-    directory_store: Database,
+    directory_database: Database,
+    lease_database: Database,
   }
 
   impl ByteStore {
     pub fn new<P: AsRef<Path>>(path: P, pool: Arc<ResettablePool>) -> Result<ByteStore, String> {
-      // 2 DBs; one for file contents, one for directories.
+      // 3 DBs; one for file contents, one for directories, one for leases.
       let env = Environment::new()
-        .set_max_dbs(2)
-        .set_map_size(16 * 1024 * 1024 * 1024)
+        .set_max_dbs(3)
+        .set_map_size(MAX_LOCAL_STORE_SIZE_BYTES)
         .open(path.as_ref())
-        .map_err(|e| format!("Error making env: {}", e.description()))?;
+        .map_err(|e| format!("Error making env: {:?}", e))?;
       let file_database = env
         .create_db(Some("files"), DatabaseFlags::empty())
         .map_err(|e| {
-          format!("Error creating/opening files database: {}", e.description())
+          format!("Error creating/opening files database: {:?}", e)
         })?;
       let directory_database = env
         .create_db(Some("directories"), DatabaseFlags::empty())
         .map_err(|e| {
-          format!(
-            "Error creating/opening directories database: {}",
-            e.description()
-          )
+          format!("Error creating/opening directories database: {:?}", e)
+        })?;
+      let lease_database = env
+        .create_db(Some("leases"), DatabaseFlags::empty())
+        .map_err(|e| {
+          format!("Error creating/opening leases database: {:?}", e)
         })?;
       Ok(ByteStore {
         inner: Arc::new(InnerStore {
-          env: env,
-          pool: pool,
-          file_store: file_database,
-          directory_store: directory_database,
+          env,
+          pool,
+          file_database,
+          directory_database,
+          lease_database,
         }),
       })
     }
-  }
 
-  impl super::ByteStore for ByteStore {
-    fn store_bytes(&self, entry_type: EntryType, bytes: Vec<u8>) -> BoxFuture<Digest, String> {
+    pub fn lease_all<Fs: Iterator<Item = Fingerprint>>(
+      &self,
+      fingerprints: Fs,
+    ) -> Result<(), String> {
+      self
+        .inner
+        .env
+        .begin_rw_txn()
+        .map_err(|err| format!("Error making lmdb transaction: {:?}", err))
+        .and_then(|mut txn| {
+          for fingerprint in fingerprints {
+            self.lease(&fingerprint, &mut txn).map_err(|e| {
+              format!("Error leasing fingerprint {}: {:?}", fingerprint, e)
+            })?;
+          }
+          Ok(txn)
+        })
+        .and_then(|txn| {
+          txn.commit().map_err(
+            |e| format!("Error writing lease: {:?}", e),
+          )
+        })
+    }
+
+    fn lease(&self, fingerprint: &Fingerprint, txn: &mut RwTransaction) -> Result<(), lmdb::Error> {
```

Comment:
Would it be useful (or just leaky?) to have lease take the expires_at value, and then have `lease_all` use one value for all entries? Would avoid some syscalls, although lease_all isn't a hot path.

Reply:
I'm cool with it. Done.